hathor/nanocontracts/storage/changes_tracker.py

```diff
@@ -0,0 +1,269 @@
+# Copyright 2023 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import itertools
+from dataclasses import dataclass
+from enum import Enum
+from types import MappingProxyType
+from typing import Any, TypeVar
+
+from typing_extensions import override
+
+from hathor.conf.settings import HATHOR_TOKEN_UID
+from hathor.nanocontracts.exception import NCInsufficientFunds, NCTokenAlreadyExists
+from hathor.nanocontracts.nc_types import NCType
+from hathor.nanocontracts.storage.contract_storage import (
+    AttrKey,
+    Balance,
+    BalanceKey,
+    MutableBalance,
+    NCContractStorage,
+)
+from hathor.nanocontracts.storage.types import _NOT_PROVIDED, DeletedKey, DeletedKeyType
+from hathor.nanocontracts.types import ContractId, TokenUid
+from hathor.transaction.token_creation_tx import TokenDescription
+
+T = TypeVar('T')
+D = TypeVar('D')
+
+
+class _NCAuthorityState(Enum):
+    """The tri-state of an authority during execution."""
+    NONE = 'none'
+    GRANTED = 'granted'
+    REVOKED = 'revoked'
+
+
+@dataclass(slots=True, kw_only=True)
+class _NCAuthorityDiff:
+    """Track the tri-state diff of each authority."""
+    mint: _NCAuthorityState = _NCAuthorityState.NONE
+    melt: _NCAuthorityState = _NCAuthorityState.NONE
+
+    def grant_mint(self) -> bool:
+        """Return whether the final mint state of this diff in granted."""
+        return self.mint is _NCAuthorityState.GRANTED
+
+    def grant_melt(self) -> bool:
+        """Return whether the final melt state of this diff in granted."""
+        return self.melt is _NCAuthorityState.GRANTED
+
+    def revoke_mint(self) -> bool:
+        """Return whether the final mint state of this diff in revoked."""
+        return self.mint is _NCAuthorityState.REVOKED
+
+    def revoke_melt(self) -> bool:
+        """Return whether the final melt state of this diff in revoked."""
+        return self.melt is _NCAuthorityState.REVOKED
+
+
+class NCChangesTracker(NCContractStorage):
```

Comment:
I think we should introduce an `NCContractStorageProtocol`, which will be implemented by both `NCContractStorage` and `NCChangesTracker`. This protocol should exclude methods like get_all_balances() and get_root_id(), as they are specific to the storage. The main goal is to ensure all storage-related methods are explicitly reimplemented, reducing the risk of accidental omissions.